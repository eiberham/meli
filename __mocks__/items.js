export const items = [{
    "id": "MLA619208970",
    "title": "Remera Lisa Algodón Jersey Peinado Premium",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/876978-MLA31035095700_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 259,
        "decimals": ""
    }
}, {
    "id": "MLA737922842",
    "title": "Remeras Personalizadas Estampada Sublimada Tu Logo",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/974505-MLA31102958584_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 199.99,
        "decimals": "99"
    }
}, {
    "id": "MLA752716977",
    "title": "Remeras De Algodon Premium Al Por Mayor",
    "condition": "new",
    "picture": "http://mla-s2-p.mlstatic.com/828346-MLA31115819250_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 260,
        "decimals": ""
    }
}, {
    "id": "MLA670757536",
    "title": "Remeras Entalladas Hombre De Importación Slim Fit!!",
    "condition": "new",
    "picture": "http://mla-s1-p.mlstatic.com/708594-MLA31133445654_062019-I.jpg",
    "free_shipping": false,
    "price": {
        "currency": "ARS",
        "amount": 249.99,
        "decimals": "99"
    }
}];