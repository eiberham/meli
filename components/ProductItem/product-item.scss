.item {
    display: flex;
    justify-content: space-around;
    align-items: center;

    &-image {
        width: 25%;
        img {
            width: 150px;
            height: 150px;
        }
    }

    &-title {
        width: 60%;
    }

    &-price {img {
        width: 150px;
        height: 150px;
    }
        width: 15%;
    }
}