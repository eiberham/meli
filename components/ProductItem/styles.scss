.item {
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-wrap: wrap;
    padding: 1.5rem 0;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);

    &:hover {
        cursor: pointer;
        box-shadow: inset 1px 0 0 #dadce0, inset -1px 0 0 #dadce0, 0 1px 2px 0 rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);
    }

    &-image {
        width: 180px;
        display:flex;
        justify-content: center;
        align-items:center;
        border-radius: 4px;
        img {
            width: 180px;
            height: 180px;
        }
    }

    &-description {
        width: 60%;
    }

    &-title {
        font-size: 18px;
    }

    &-price {
        font-size: 24px;
    }

    &-decimals {
        line-height: 1.8;
        font-size: 14px;
        position: relative;
        top: -8px;
    }

    &-condition {
        font-size: 12px;
        padding: 0 2.5rem;
        width: 150px;
        text-align: center;
    }
}

@media screen and (max-width: 768px) {
    .item {
        /* display: flex; */
        justify-content: center;
        /* align-items: center; */
        text-align: center;
    }
}