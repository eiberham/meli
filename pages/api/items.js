export default (req, res) => {
    console.log("el objeto req negro: ", req);
    res.send("Todo bien papu");
}