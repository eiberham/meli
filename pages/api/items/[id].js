import redis from 'redis';
import axios from 'axios';
import _ from 'lodash';
import meli from '../../../services/meli';

export default (req, res) => {
    const client = redis.createClient();
    let id = req.query.id;

    client.get(id, async (err, result) => {
        if (result) {
            res.send(result);
        } else {
            axios.all([
                meli.get(`/items/${id}`),
                meli.get(`/items/${id}/description`)
            ])
                .then(axios.spread(async (first, second) => {
                    let response = {...first.data, ...second.data};

                    let data = {
                        ..._.pick(response, ['id', 'title', 'condition', 'sold_quantity']),
                        picture: _.get(response, 'thumbnail'),
                        free_shipping: _.get(response, 'shipping.free_shipping'),
                        description: _.get(response, 'plain_text'),
                        price: {
                            currency: _.get(response, 'currency_id'),
                            amount: _.get(response, 'price'),
                            decimals: _.toString(_.split(_.get(response, 'price'), '.')[1])
                        }
                    };

                    const {data:{path_from_root}} = await meli.get(`/categories/${response.category_id}`);
                    let categories = _.map(path_from_root, item => {
                        return item.name
                    });

                    let output = {
                        author: { name: 'Abraham', lastname: 'Cedeno'},
                        item: data,
                        categories
                    };

                    client.setex(id, 3600, JSON.stringify(output));

                    res.status(200).json(output);
                }))
                .catch(({response:{data}}) => {
                    if(data.status === 404){
                        return res.status(404).json({
                            "status": 404,
                            "message": data.message
                        });
                    }
                });
        }
    });
}