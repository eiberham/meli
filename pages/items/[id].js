import fetch from 'isomorphic-unfetch';
import Layout from '../../components/Layout';
import Breadcrumb from '../../components/Breadcrumb';
import ProductDetail from '../../components/ProductDetail';

const Item = props => {
    const { item, categories } = props;

    return (
        <Layout>
            <Breadcrumb links={categories} />
            <ProductDetail item={item} />
        </Layout>
    )
}

Item.getInitialProps = async context => {
    const {
        query: { id },
    } = context;
    
    const response = await fetch(`http://localhost:3000/api/items/${id}`);
    const data = await response.json();
    return data;
}

export default Item;