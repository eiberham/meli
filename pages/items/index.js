import { Fragment } from 'react';
import fetch from 'isomorphic-unfetch';
import Layout from '../../components/Layout';
import Breadcrumb from '../../components/Breadcrumb';
import ProductList from '../../components/ProductList';
import Loading from '../../components/Loading';
import { withRouter } from 'next/router';

const Items = props => {
    const { items, router, categories } = props;

    const onItemSelect = (id) => {
        router.push(`/items/${id}`)
    }

    return (
        <Layout>
            { (items) ? (
                <Fragment>
                    <Breadcrumb links={categories} />
                    <ProductList items={items} onItemSelect={onItemSelect}/>
                </Fragment>
            ) : (
                <Loading size="massive" show />
            )}
        </Layout>
    );
};

Items.getInitialProps = async (context) => {
    const {
        query: { search },
    } = context;
    
    const response = await fetch(`http://localhost:3000/api/items?q=${search}`);
    const data = await response.json();
    return data;
}

export default withRouter(Items);