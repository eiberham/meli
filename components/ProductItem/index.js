import React from 'react';
import './styles.scss';
import PropTypes from 'prop-types';
import items from '../../pages/items';

const ProductItem = ({product, onProductSelect}) => {
    return (
        <React.Fragment>
            <div className="item" onClick={onProductSelect}>
                <div className="item-image">
                    <img src={product.picture} alt={product.title} />
                </div>
                
                <div className="item-description">
                    <div className="item-price">
                        $ {(parseInt(product.price.amount)).toLocaleString()}
                        <span className="item-decimals">{product.price.decimals}</span>
                    </div>
                    <div className="item-title">{product.title}</div>
                </div>

                <div className="item-condition">
                    {product.condition === 'new' ? (
                        <span>Nuevo</span>
                    ) : ( 
                        <span>Usado</span>
                    )}
                </div>
                
            </div>
        </React.Fragment>
    );
};

ProductItem.propTypes = {
    product: PropTypes.shape({
        id: PropTypes.string,
        title: PropTypes.string,
        condition: PropTypes.string,
        free_shipping: PropTypes.bool,
        picture: PropTypes.string,
        price: PropTypes.shape({
            amount: PropTypes.number,
            currency: PropTypes.string,
            decimals: PropTypes.string
        })
    }).isRequired,
    onProductSelect: PropTypes.func
}

export default ProductItem;