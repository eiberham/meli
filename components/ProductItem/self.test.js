import React from 'react';
import { items } from '../../__mocks__/items';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ProductItem from './index';

describe('ProductItem', () => {
    it('matches snapshot', () => {
        const onProductSelect = jest.fn();
        const product = items[0];
        const component = renderer.create(<ProductItem product={product} onProductSelect={onProductSelect} />)
        .toJSON();
        expect(component).toMatchSnapshot();
    })
})