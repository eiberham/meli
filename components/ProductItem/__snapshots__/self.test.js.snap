// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem matches snapshot 1`] = `
<div
  className="item"
  onClick={[MockFunction]}
>
  <div
    className="item-image"
  >
    <img
      alt="Remera Lisa Algodón Jersey Peinado Premium"
      src="http://mla-s1-p.mlstatic.com/876978-MLA31035095700_062019-I.jpg"
    />
  </div>
  <div
    className="item-description"
  >
    <div
      className="item-price"
    >
      $ 
      259
      <span
        className="item-decimals"
      >
        
      </span>
    </div>
    <div
      className="item-title"
    >
      Remera Lisa Algodón Jersey Peinado Premium
    </div>
  </div>
  <div
    className="item-condition"
  >
    <span>
      Nuevo
    </span>
  </div>
</div>
`;
